File: backend/chameleon-core/src/chameleon/core/models/graph.py
"""Graph ORM —— P18.1 工作流引擎持久层

三张表：
- graphs        图声明（spec 落 JSONB）
- graph_runs    一次执行的总记录
- graph_node_runs 每节点执行的子记录，request_id 与 call_logs 串联
"""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import (
    JSON,
    BigInteger,
    Boolean,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column

from chameleon.core.models.base import Base, SoftDeleteMixin, TimestampMixin, snowflake_pk
from chameleon.core.models.workspace import WorkspaceScopedMixin


class Graph(Base, TimestampMixin, SoftDeleteMixin, WorkspaceScopedMixin):
    """工作流图声明"""

    __tablename__ = "graphs"

    id: Mapped[int] = snowflake_pk()
    graph_key: Mapped[str] = mapped_column(
        String(64), unique=True, nullable=False
    )
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    schema_version: Mapped[int] = mapped_column(
        Integer, nullable=False, default=1
    )
    # GraphSpec.model_dump() 落 JSONB（draft 版本，可改）
    spec: Mapped[dict] = mapped_column(JSON, nullable=False)
    # P22.3：published 版本快照（freeze；admin publish 时从 spec 拷贝过来）
    published_spec: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    published_version: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0, server_default="0"
    )
    published_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    enabled: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True
    )


class GraphRun(Base):
    """一次完整执行的记录"""

    __tablename__ = "graph_runs"

    id: Mapped[int] = snowflake_pk()
    graph_id: Mapped[int] = mapped_column(
        BigInteger,
        ForeignKey("graphs.id", ondelete="CASCADE"),
        nullable=False,
    )
    # 与 call_logs.request_id 一致，串联 trace tree
    request_id: Mapped[str] = mapped_column(
        String(64), unique=True, nullable=False
    )
    status: Mapped[str] = mapped_column(
        String(16), nullable=False, default="pending"
    )  # pending/running/success/failed/cancelled
    input: Mapped[dict] = mapped_column(JSON, nullable=False)
    output: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    error: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    started_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    finished_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    duration_ms: Mapped[int | None] = mapped_column(Integer, nullable=True)
    node_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    __table_args__ = (Index("ix_graph_runs_graph", "graph_id", "created_at"),)


class GraphNodeRun(Base):
    """单节点执行记录"""

    __tablename__ = "graph_node_runs"

    id: Mapped[int] = snowflake_pk()
    graph_run_id: Mapped[int] = mapped_column(
        BigInteger,
        ForeignKey("graph_runs.id", ondelete="CASCADE"),
        nullable=False,
    )
    node_id: Mapped[str] = mapped_column(String(64), nullable=False)
    node_type: Mapped[str] = mapped_column(String(32), nullable=False)
    status: Mapped[str] = mapped_column(
        String(16), nullable=False, default="pending"
    )
    input: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    output: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    error: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    # 写入 call_logs.parent_id 实现 trace 串联（PR #21 起填充）
    request_id: Mapped[str | None] = mapped_column(
        String(64), nullable=True
    )
    started_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    finished_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    duration_ms: Mapped[int | None] = mapped_column(Integer, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    __table_args__ = (
        Index("ix_graph_node_runs_run", "graph_run_id"),
        Index("ix_graph_node_runs_rid", "request_id"),
    )

File: frontend/src/system/graphs/types/graph.ts
import type { EntityId } from '@/core/types/api';

export type GraphNodeType =
  | 'noop'
  | 'start'
  | 'end'
  | 'llm'
  | 'kb'
  | 'tool'
  | 'if_else'
  | 'agent_debate'
  | 'iteration'
  | 'parallel'
  | 'human_input'
  | 'template'
  | 'answer'
  | 'http'
  | 'aggregator'
  | 'assign'
  | 'code'
  | 'classifier';

export interface NodeSpec {
  id: string;
  type: GraphNodeType;
  name?: string | null;
  data?: Record<string, unknown>;
  position?: { x: number; y: number } | null;
}

export interface EdgeSpec {
  id: string;
  source: string;
  target: string;
  source_handle?: string | null;
}

export interface GraphSpec {
  nodes: NodeSpec[];
  edges: EdgeSpec[];
}

/** 工作流形态：对话型（聊天 I/O，可发布为智能体）/ 流程型（一次性管线） */
export type GraphKind = 'chatflow' | 'workflow';

export interface GraphItem {
  id: EntityId;
  graph_key: string;
  name: string;
  description: string | null;
  kind: GraphKind;
  schema_version: number;
  enabled: boolean;
  /** P22.3：已发布版本号（0 = 从未发布） */
  published_version?: number;
  /** P22.3：最近一次发布时间 */
  published_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface GraphDetail extends GraphItem {
  spec: GraphSpec;
  /** P22.3：已发布的 spec 快照（freeze） */
  published_spec?: GraphSpec | null;
}

/** Web App / 嵌入信息（基于 embed_config） */
export interface WebAppInfo {
  id: EntityId;
  embed_key: string;
  agent_key: string;
  name: string;
  description: string | null;
  ui_config: Record<string, unknown>;
  behavior: Record<string, unknown>;
  enabled: boolean;
}

/** 智能体级 API 密钥（仅对该 agent 有效，区别于全局应用密钥） */
export interface AgentApiKey {
  id: EntityId;
  name: string;
  key_prefix: string;
  /** 明文 key（留存，支持重复复制；老数据为 null 只能看前缀） */
  plain_key: string | null;
  agent_key: string | null;
  created_at: string;
  last_used_at: string | null;
  revoked_at: string | null;
}

export type AgentApiKeyCreated = AgentApiKey;

export interface NodeRunItem {
  node_id: string;
  node_type: string;
  status: 'pending' | 'running' | 'success' | 'failed' | 'skipped';
  input?: unknown;
  output?: unknown;
  error?: { type: string; message: string } | null;
  duration_ms: number;
}

export interface TestRunResult {
  status: 'success' | 'failed';
  output?: unknown;
  error?: { type: string; message: string } | null;
  duration_ms: number;
  node_runs: NodeRunItem[];
}

export interface GraphRunItem {
  id: EntityId;
  graph_id: EntityId;
  request_id: string;
  session_id: string | null;
  status: 'pending' | 'running' | 'success' | 'failed' | 'cancelled' | 'paused';
  duration_ms: number | null;
  node_count: number | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
}

export interface GraphRunDetail extends GraphRunItem {
  input?: unknown;
  output?: unknown;
  error?: { type: string; message: string } | null;
  node_runs: NodeRunItem[];
}

// ── 调试运行视图（编辑器内）─────────────────────────────────

/** 单节点在一次运行中的投影：驱动 canvas 染色 + inspector 结果区 */
export interface NodeRunView {
  status: NodeRunItem['status'];
  input?: unknown;
  output?: unknown;
  error?: { type: string; message: string } | null;
  duration_ms?: number;
  /** 流式累积的 token 文本（LLM 节点 delta，可选） */
  streamText?: string;
}

// ── test-run/stream SSE chunk（镜像后端 SSEEventKind 的 graph.* 成员）──
// wire 形状：扁平 dict { "<kind>": payload }，与 core/api/sse.py 对齐。

export interface GraphNodeEventPayload {
  node_id: string;
  node_type?: string | null;
  name?: string | null;
  status?: string | null; // running / success / failed
  duration_ms?: number | null;
  output?: unknown;
  error?: { type: string; message: string } | null;
}

export interface GraphFinishedPayload {
  status: 'success' | 'failed' | string;
  duration_ms?: number;
  node_count?: number;
  output?: unknown;
  error?: { type: string; message: string } | null;
}

/** 任一 chunk 只含一个 kind 键；前端按 key 分派。 */
export type GraphStreamChunk =
  | { 'graph.started': Record<string, unknown> }
  | { 'graph.node.started': GraphNodeEventPayload }
  | { 'graph.node.delta': { node_id: string; delta: string } }
  | { 'graph.node.finished': GraphNodeEventPayload }
  | { 'graph.node.failed': GraphNodeEventPayload }
  | { 'graph.finished': GraphFinishedPayload };

// ── 对话调试（把 draft 当可对话 agent 跑）SSE chunk ──────────
// 后端 /chat/stream 形状：{ type, data }（GraphProvider StreamEvent 镜像）

export interface GraphChatChunk {
  type: 'delta' | 'step' | 'done' | 'error' | 'citation';
  data: {
    text?: string; // delta
    name?: string; // step
    status?: string; // step
    duration_ms?: number | null;
    answer?: string; // done
    message?: string; // error
    source?: string; // citation
    snippet?: string; // citation
    score?: number | null; // citation
    [k: string]: unknown;
  };
}
